src/coreclr/src/utilcode/pedecoder.cpp

```diff
@@ -2319,22 +2319,70 @@ READYTORUN_HEADER * PEDecoder::FindReadyToRunHeader() const
     }
     CONTRACTL_END;
 
-    IMAGE_DATA_DIRECTORY *pDir = &GetCorHeader()->ManagedNativeHeader;
-
-    if (VAL32(pDir->Size) >= sizeof(READYTORUN_HEADER) && CheckDirectory(pDir))
+    if (HasCorHeader())
     {
-        PTR_READYTORUN_HEADER pHeader = PTR_READYTORUN_HEADER((TADDR)GetDirectoryData(pDir));
-        if (pHeader->Signature == READYTORUN_SIGNATURE)
+        IMAGE_DATA_DIRECTORY* pDir = &GetCorHeader()->ManagedNativeHeader;
+
+        if (VAL32(pDir->Size) >= sizeof(READYTORUN_HEADER) && CheckDirectory(pDir))
         {
-            const_cast<PEDecoder *>(this)->m_pReadyToRunHeader = pHeader;
-            return pHeader;
+            PTR_READYTORUN_HEADER pHeader = PTR_READYTORUN_HEADER((TADDR)GetDirectoryData(pDir));
+            if (pHeader->Signature == READYTORUN_SIGNATURE)
+            {
+                const_cast<PEDecoder*>(this)->m_pReadyToRunHeader = pHeader;
+                return pHeader;
+            }
         }
     }
+    
+    READYTORUN_HEADER *nativeReadyToRunHeader = FindNativeReadyToRunHeader();
+    if (nativeReadyToRunHeader != NULL)
+    {
+        return nativeReadyToRunHeader;
+    }
 
     const_cast<PEDecoder *>(this)->m_flags |= FLAG_HAS_NO_READYTORUN_HEADER;
     return NULL;
 }
 
+READYTORUN_HEADER *PEDecoder::FindNativeReadyToRunHeader() const
+{
+#ifndef DACCESS_COMPILE
+    // Get the export directory entry
+    PIMAGE_DATA_DIRECTORY exportDirectoryEntry = GetDirectoryEntry(IMAGE_DIRECTORY_ENTRY_EXPORT);
+    if (exportDirectoryEntry->VirtualAddress == 0 || exportDirectoryEntry->Size == 0)
+    {
+        return NULL;
+    }
+    
+    const uint8_t *imageBase = (const uint8_t *)GetBase();
+    const uint8_t *exportDirectory = &imageBase[exportDirectoryEntry->VirtualAddress];
+    
+    // TODO: endianness?
+    // +0x18: number of name pointers
+    int namePointerCount = *(const int32_t *)&exportDirectory[0x18];
+    // +0x1C: export address table RVA
+    int addressTableRVA = *(const int32_t *)&exportDirectory[0x1C];
+    // +0x20: name pointer RVA
+    int namePointersRVA = *(const int32_t *)&exportDirectory[0x20];
+
+    for (int nameIndex = 0; nameIndex < namePointerCount; nameIndex++)
+    {
+        int namePointerRVA = *(int32_t *)&imageBase[namePointersRVA + sizeof(int32_t) * nameIndex];
```

Comment:
Is it possible that a corrupted PE file will make this go outside the bounds of imageBase?  Is it possible that the index expression might overflow?

Reply:
Fixed by means of using a new helper ReadUInt32 that includes the range check.